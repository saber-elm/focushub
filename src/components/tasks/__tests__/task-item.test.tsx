import { fireEvent, render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import { Task } from '@/types/task'
import TaskItem from '../task-item'
import { axe, toHaveNoViolations } from 'jest-axe'

const mockTask: Task = {
  id: '1',
  title: 'Test Task',
  completed: false,
}

describe('TaskItem', () => {
  it('renders task title', () => {
    render(<TaskItem task={mockTask} onToggle={() => {}} onDelete={() => {}} />)
    expect(screen.getByText(/Test Task/i)).toBeInTheDocument()
  })

  it('Calls onToggle when toggle is clicked', () => {
    const onToggle = vi.fn()
    render(<TaskItem task={mockTask} onToggle={onToggle} onDelete={() => {}} />)

    const toggleInput = screen.getByTestId('task-toggle')
    fireEvent.click(toggleInput)

    expect(onToggle).toHaveBeenCalled()
  })

  it('Calls onDelete when delete button is clicked', () => {
    const onDelete = vi.fn()
    const { getByText } = render(
      <TaskItem task={mockTask} onToggle={() => {}} onDelete={onDelete} />
    )

    const deleteButton = getByText(/❌ Delete/i)
    deleteButton.click()

    expect(onDelete).toHaveBeenCalledWith(mockTask.id)
  })

  it('displays completed task with line-through style', () => {
    const completedTask = { ...mockTask, completed: true }

    const { getByText } = render(
      <TaskItem task={completedTask} onToggle={() => {}} onDelete={() => {}} />
    )
    const taskText = getByText(/Test Task/i)
    expect(taskText).toHaveClass('line-through')
  })

  it('renders completed task correctly', () => {
    const completedTask: Task = {
      id: '2',
      title: 'Complete Task',
      completed: true,
    }

    const { getByText, getByTestId } = render(
      <TaskItem task={completedTask} onToggle={() => {}} onDelete={() => {}} />
    )

    const title = getByText(/Complete Task/i)
    const checkbox = getByTestId('task-toggle')
    const label = getByText(/Completed/i)

    expect(title).toHaveClass('line-through')
    expect(checkbox).toBeChecked()
    expect(label).toHaveTextContent('Completed')
  })

  expect.extend(toHaveNoViolations)
  it('should have no accessibility violations', async () => {
    const { container } = render(
      <TaskItem task={mockTask} onToggle={() => {}} onDelete={() => {}} />
    )
    const result = await axe(container)
    expect(result).toHaveNoViolations()
  })

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <TaskItem task={mockTask} onToggle={() => {}} onDelete={() => {}} />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})