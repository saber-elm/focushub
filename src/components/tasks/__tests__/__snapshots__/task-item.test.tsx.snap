// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskItem > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between border px-4 py-2 rounded"
  >
    <span
      class="max-w-xl wrap-break-word min-w-1"
    >
      Test Task
    </span>
    <div
      class="flex items-center gap-4"
    >
      <div
        class="space-x-1"
      >
        <input
          class="text-blue-600 hover:underline"
          data-testid="task-toggle"
          id="task-1"
          name="status"
          type="checkbox"
        />
        <label
          class="hover:underline"
          for="task-1"
        >
          inComplete
        </label>
      </div>
      <button
        class="text-red-500 hover:underline"
      >
        ‚ùå Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
